diff --git a/meson.build b/meson.build
index 32ef797..282d9ab 100644
--- a/meson.build
+++ b/meson.build
@@ -45,9 +45,11 @@ cflags = cc.get_supported_arguments(
 add_project_arguments(cflags, language: 'c')
 
 # Dependencies
-pkgconfig   = import('pkgconfig')
-dep_gudev   = dependency('gudev-1.0')
-dep_glib    = dependency('glib-2.0')
+pkgconfig      = import('pkgconfig')
+dep_gudev      = dependency('gudev-1.0')
+dep_glib       = dependency('glib-2.0')
+build_dep_gudev = dependency('gudev-1.0', native: true)
+build_dep_glib  = dependency('glib-2.0', native: true)
 
 includes_include = include_directories('include')
 includes_src = include_directories('libwacom')
@@ -68,6 +70,10 @@ deps_libwacom = [
 	    dep_glib
 ]
 
+build_deps_libwacom = [
+	    build_dep_gudev,
+	    build_dep_glib
+]
 inc_libwacom = [
 	     includes_include,
 	     includes_src,
@@ -87,7 +93,22 @@ lib_libwacom = shared_library('wacom',
 				'-DDATADIR="@0@"'.format(dir_data),
 			      ],
 			      install: true)
+build_lib_libwacom = shared_library('wacom-for-generator',
+			      src_libwacom,
+			      include_directories: inc_libwacom,
+			      dependencies: build_deps_libwacom,
+			      version: libwacom_so_version,
+			      link_args: version_flag,
+			      link_depends: mapfile,
+			      c_args: [
+				'-DG_LOG_DOMAIN="@0@"'.format(meson.project_name()),
+				'-DDATADIR="@0@"'.format(dir_data),
+			      ],
+			      install: false,
+			      native: true)
+
 dep_libwacom = declare_dependency(link_with: lib_libwacom)
+build_dep_libwacom = declare_dependency(link_with: build_lib_libwacom)
 
 install_headers('libwacom/libwacom.h', subdir: 'libwacom-1.0/libwacom')
 
@@ -135,10 +156,11 @@ tools_cflags = ['-DTOPSRCDIR="@0@"'.format(meson.source_root())]
 
 gen_hwdb = executable('generate-hwdb',
 		      'tools/generate-hwdb.c',
-		      dependencies: [dep_libwacom, dep_glib],
+		      dependencies: [build_dep_libwacom, build_dep_glib],
 		      include_directories: [includes_src],
 		      c_args: tools_cflags,
-		      install: false)
+		      install: false,
+		      native: true)
 custom_target('hwdb',
 	      command: gen_hwdb,
 	      capture: true,
